In [1]:
#載入所需函示庫

from matplotlib import pyplot as plt
%matplotlib inline
# 設定圖形大小; DPI越大圖越大
plt.rcParams["figure.dpi"] = 120

import pandas as pd
import numpy as np
import scipy.stats as stats
import statsmodels.api as sm
import statsmodels.stats.api as sms
import statsmodels.formula.api as smf
import statsmodels.stats.multicomp as smm
import statsmodels
import scipy

### 評分標準

每題16% : 
- H0H1 2% (15.12 15.19沒直接寫出各組期望機率-1%，15.36 15.43沒寫independent-1%)
- 計算 5% （錯一個值-2%）
- RejectH0 2% (pvalue/critical value判斷1%，reject or not 1%)
- Conclusion 2%  
- Rules of 5 5% (若只有display freq_e -2%)


### 15.12 

Let p1, p2, p3, p4 be the expected proportions of account receivable for group1(0-14), group2(15-29), group3(30-59), group4(>60).  

H0: p1 = .72, p2 = .15, p3 = .10, p4 = .03  
H1: At least one pi is not equal to its specified value.  

Alpha = 0.05

In [2]:
data12 = pd.read_excel("Xr15-12.xlsx")
display(data12)

Accounts Receivable
0                      3
1                      1
2                      1
3                      1
4                      1
..                   ...
245                    1
246                    2
247                    1
248                    1
249                    1

[250 rows x 1 columns]

In [3]:
nobs = len(data12)
alpha = 0.05
prob_e = np.array([0.72, 0.15, 0.10, 0.03])
freq_e = nobs * prob_e
print(freq_e)

freq = data12.apply(pd.value_counts)

#注意value counts 後的類別排序
print(freq,'\n')
freq_o = freq.sort_index().iloc[:,0].values

if np.sum(freq_e < 5) > 0:
    print("Rule of five is not met. \n")
else:
    print("Rule of five is met. \n")
    stat, pvalue = stats.chisquare(freq_o, freq_e)
    print("Chi-squared test: statistics = %0.4f, p-value = %0.04f" % (stat, pvalue))

    df = freq_o.shape[0]-1
    crit_value = stats.chi2.ppf(1 - alpha, df)
    print("Critical value = %0.4f (degree of freedom = %d)" % (crit_value, df))

[180.   37.5  25.    7.5]
   Accounts Receivable
1                  159
3                   47
2                   28
4                   16 

Rule of five is met. 

Chi-squared test: statistics = 33.8500, p-value = 0.0000
Critical value = 7.8147 (degree of freedom = 3)


Since p-value = 0 < alpha = 0.05, we can reject the null hypothesis at a significance level of 0.05.  
There is enough evidence to infer that the aging schedule has changed.

### 15.19 

Let pi be the expected percentages of monthly number of tornadoes for the ith month in a year.  

H0: p1 = .024, p2 = .030, p3 = .067, p4 = .169, p5 = .212, P6 = .174, p7 = .080, p8 = .057, p9 = .059, p10 = .049, p11 = .049, p12 = .028  
H1: At least one pi is not equal to its specified value.


Alpha = 0.05

In [4]:
data19 = pd.read_excel("Xr15-19.xlsx")
print('Head of dataset:')
display(data19)

Head of dataset:


Month  Number of tornadoes
0     January                   26
1    February                    2
2       March                   13
3       April                  185
4         May                  414
5        June                  184
6        July                  122
7      August                   44
8   September                   24
9     October                   57
10   November                   95
11   December                   93

In [5]:
nobs = sum(data19.iloc[:,1].values)
alpha = 0.05

prob = [0.024, 0.030, 0.067, 0.169, 0.212, 0.174, \
        0.080, 0.057, 0.059, 0.049, 0.049, 0.028]
prob_e = np.array(prob)
freq_e = nobs * prob_e
print(freq_e)

freq_o = data19.iloc[:,1].values


if np.sum(freq_e < 5) > 0:
    print("Rule of five is not met. \n")
else:
    print("Rule of five is met. \n")
    stat, pvalue = stats.chisquare(freq_o, freq_e)
    print("Chi-squared test: statistics = %0.4f, p-value = %0.10f" % (stat, pvalue))

    df = freq_o.shape[0]-1
    crit_value = stats.chi2.ppf(1 - alpha, df)
    print("Critical value = %0.4f (degree of freedom = %d)" % (crit_value, df))

[ 30.216  37.77   84.353 212.771 266.908 219.066 100.72   71.763  74.281
  61.691  61.691  35.252]
Rule of five is met. 

Chi-squared test: statistics = 347.3334, p-value = 0.0000000000
Critical value = 19.6751 (degree of freedom = 11)


Since p-value = 0 < alpha = 0.05, we can reject the null hypothesis at a significance level of 0.05.  
There is enough evidence to infer that the percentages have changed.

### 15.36 

H0: The two variables (remedy/placebo and side effects) are independent.  
H1: The two variables are dependent.  

Alpha = 0.05

In [6]:
data36 = pd.read_excel("Xr15-36.xlsx")
print('Head of dataset:')
display(data36)

Head of dataset:


Cold Remedy  Placebo
0              4        4
1              4        3
2              4        4
3              4        4
4              4        4
..           ...      ...
245            4        4
246            4        4
247            4        4
248            4        4
249            3        4

[250 rows x 2 columns]

In [7]:
# 列出Contingency Table
freq_o = [data36.apply(pd.value_counts).sort_index().iloc[:,0].values,\
          data36.apply(pd.value_counts).sort_index().iloc[:,1].values]
print("Contingency Table:")
display(freq_o)

Contingency Table:


[array([ 19,  23,  14, 194]), array([ 17,  18,  16, 199])]

In [8]:
chi2, p, dof, ex =stats.chi2_contingency(freq_o, correction = False)
print("Chi-squared Statistics = %0.4f" % chi2)
print("Degree of freedom = %0.4f" % dof)
print("p-value = %0.4f" % p)

print("\nExpected Frequency:")
print(ex)
if np.sum(ex < 5) > 0:
    print("Rule of five is not met. ")
else:
    print("Rule of five is met. ")


Chi-squared Statistics = 0.9178
Degree of freedom = 3.0000
p-value = 0.8211

Expected Frequency:
[[ 18.   20.5  15.  196.5]
 [ 18.   20.5  15.  196.5]]
Rule of five is met. 


Since p-value = 0.8211 > alpha = 0.05. We can not reject the null hypothesis at a significance level of 0.05.  
There is not enough evidence to infer that the reported side effects differ, which means the two variables (remedy/placebo and side effects) are independent.

### 15.43 

H0: The two variables (year and weapon) are independent.  
H1: The two variables are dependent.  

Alpha = 0.05

In [9]:
data43 = pd.read_excel("Xr15-43.xlsx")
print('Head of dataset:')
display(data43)

Head of dataset:


Year 2000  Year 2005  Year 2010
0          3.0          1        4.0
1          4.0          4        4.0
2          1.0          3        2.0
3          1.0          2        4.0
4          4.0          1        1.0
..         ...        ...        ...
412        NaN          1        NaN
413        NaN          4        NaN
414        NaN          4        NaN
415        NaN          4        NaN
416        NaN          4        NaN

[417 rows x 3 columns]

In [10]:
# 列出Contingency Table
freq_o = [data43.apply(pd.value_counts).sort_index().iloc[:,0].values,\
          data43.apply(pd.value_counts).sort_index().iloc[:,1].values,\
          data43.apply(pd.value_counts).sort_index().iloc[:,2].values]

print("Contingency Table:")
display(freq_o)

Contingency Table:


[array([161,  36,  53, 159]),
 array([175,  37,  39, 166]),
 array([131,  24,  27, 126])]

In [11]:
chi2, p, dof, ex =stats.chi2_contingency(freq_o, correction = False)
print("Chi-squared Statistics = %0.4f" % chi2)
print("Degree of freedom = %0.4f" % dof)
print("p-value = %0.4f" % p)

print("\nExpected Frequency:")
print(ex)
if np.sum(ex < 5) > 0:
    print("Rule of five is not met. ")
else:
    print("Rule of five is met. ")

Chi-squared Statistics = 4.7588
Degree of freedom = 6.0000
p-value = 0.5751

Expected Frequency:
[[168.4329806   34.98500882  42.91975309 162.6622575 ]
 [171.72751323  35.66931217  43.75925926 165.84391534]
 [126.83950617  26.34567901  32.32098765 122.49382716]]
Rule of five is met. 


Since p-value = 0.5751 > alpha = 0.05. We can not reject the null hypothesis at a significance level of 0.05.  
There is not sufficient evidence to infer that the frequency of the use of weapons in robberies differed over the three years, which means the two variables are independent.

### 15.80 

H0:The data are normally distributed.  
H1:The data are not normally distributed.  

Alpha = 0.10

In [12]:
freq_o = [6, 27, 14, 3]
nobs = sum(freq_o)
alpha = 0.10

prob = []
prob.append(stats.norm.cdf(-1))
prob.append(stats.norm.cdf(0)-stats.norm.cdf(-1))
prob.append(stats.norm.cdf(1)-stats.norm.cdf(0))
prob.append(1 - stats.norm.cdf(1))
#print(prob)

prob_e = np.array(prob)
freq_e = nobs * prob_e
#print(freq_e)


if np.sum(freq_e < 5) > 0:
    print("Rule of five is not met. \n")
else:
    print("Rule of five is met. \n")
    stat, pvalue = stats.chisquare(freq_o, freq_e,  ddof = 2)
    print("Chi-squared test: statistics = %0.4f, p-value = %0.10f" % (stat, pvalue))

    df = len(freq_o)-1-2
    crit_value = stats.chi2.ppf(1 - alpha, df)
    print("Critical value = %0.4f (degree of freedom = %d)" % (crit_value, df))

Rule of five is met. 

Chi-squared test: statistics = 9.8701, p-value = 0.0016798788
Critical value = 2.7055 (degree of freedom = 1)


Since p-value = 0.0017 < alpha = 0.10. We can reject the null hypothesis at a significance level of 0.10.  
There is sufficient evidence to infer that the data are not normally distributed.

### 15.85  

H0:Matched pairs differences of sales are normally distributed.  
H1:Matched pairs differences of sales are not normally distributed.  


Alpha = 0.10

In [13]:
data85 = pd.read_excel("Xr13-115.xlsx")
print('Head of dataset:')
display(data85)

Head of dataset:


Method 1  Method 2
0       8.8       9.2
1       9.6       9.4
2       8.4       8.9
3       9.0       9.6
4       8.3       9.7
5       9.2       8.4
6       9.0       8.8
7       8.7       8.9
8       8.5       9.0
9       9.4       9.7

To fulfill rule of 5, we devide the data into two parts. However, this causes the problem of degree of freedom falls to -1, which is not calculatable.